### PR TITLE
Support for: When all weights are 0, try possible unsafe resources

### PR DESCRIPTION
**Pull request by henriquegemignani**

### CODE REVIEW

**Comment by JeffGainsNGames** [comment]
data https://docs.google.com/spreadsheets/d/1nx6r5-CraL70WmjrmqhN8ZNWE_K3AzKOQf1QgIylz1M/edit?usp=sharing

**Review by Miepee — COMMENTED** [review]
IMO this should either have an experimental flag, or we need more data from games to properly review it

Ideally former one is done first so that this isn't stuck in PR hell, since evaluating the result this has on games will probably take quite a while.

**Comment by Miepee** [comment]
The second commit feels like it should've been its own PR. If you wanna keep it in here, it should be excluded from the flag tho since it's an optimization without any side effects.

**Comment by henriquegemignani** [comment]
This has four changes:

1. The short circuit for 1 action.
2. The refactoring to let the rest be simpler.
3. The additional fallback when weights are 0.
4. Applying offsets and mulitpliers only at the end, so that both "weight are 0" fallbacks can happen.

I left 1 in the PR because it would explicitly conflict with this PR anyway.
1 and 2 doesn't need settings, 3 is doable but for 4 it'd be an ugly mess which I particularly don't want to.

However I don't see this change as one that have unknown results. It explicitly makes the actions chosen when stuck to be more likely to unlock the generator.

**Comment by Miepee** [comment]
> 1 and 2 doesn't need settings, 3 is doable but for 4 it'd be an ugly mess which I particularly don't want to

Yeah 3 is the one i meant for a setting

> However I don't see this change as one that have unknown results. 

I only looked at it for prime and fusion. For me, prime seemed decent, but for fusion the longer i looked at it, the more negative the result seemed to me. I'm not quite sure without looking much deeper into it whether thats just how fusion is, or whether fusions young unmatured db is at fault there where the pr just exposes that unmatureness. So i'd at least like for the echoes/dread people to judge the effects on their game too.

**Comment by henriquegemignani** [comment]
Made the new fallback weighting be configurable, but not adding it to presets - it's always off right now. Still has some weighting changes for the backup weights due to code changes though.

File: randovania/generator/filler/retcon.py [hunk 1]
```diff
@@ -93,6 +94,52 @@ def _calculate_weight(player: PlayerState) -> float:
...
+    state = base_reach.state
+    multiplier = 1.0
+    offset = 0.0
+
+    resources, pickups = action.split_pickups()
+
+    if resources:
+        for resource in resources:
+            state = state.act_on_node(resource)
+        multiplier *= action_weights.DANGEROUS_ACTION_MULTIPLIER
```
[the remark this answers is not in the document]
Comment (henriquegemignani) [reply randovania/generator/filler/retcon.py:128]: The only kind of resource that ends up as an action are dangerous resources.

[the remark this answers is not in the document]
Comment (Miepee) [reply randovania/generator/filler/retcon.py:128]: ahhh ok

**Review by Miepee — APPROVED** [review]

**Pull request merged**